src/serialize.hpp

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 Chia Network Inc
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//    http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SRC_SERIALIZE_HPP_
+#define SRC_SERIALIZE_HPP_
+
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+template<typename Type> class Serializable{
```

Comment:
it seems risky to me, to default to making everything serializable by just saving its bytes.
I think this should be constrained to specific types. Never pointers or references for instance.

I think you should add something like this:

```
template<typename Type, std::enable_if_t<std::is_trivial_v<Type>, int> = 0>
class Serializable {
```

Reply:
Right. The problem with `static_assert` is that you no longer can detect whether `Serializable` is specialized or not.

you could do it like this:
```
template <typename T, typename Dummy = int>
struct Serializable {
    static bool const is_specialized = false;
};

template <typename T>
struct Serializable<T, typename std::enable_if_t<std::is_trivial_v<T>, int>> {
    static bool const is_specialized = true;
    // ...
};

template <>
struct A<std::string, int> {
    static bool const is_specialized = true;
    // ...
};
```